pages/api/labs/games/multiplication-connect/Index.test.ts

```diff
@@ -0,0 +1,169 @@
+        expect(calculateWinner(grid, isPlayerOne)).toEqual("PLAYERONE (vertical) Four in a row!");
+    });
+    test("PlayerTwo vertical win", () => {
+        let grid = createGrid();
+        let isPlayerOne = false;
+        grid[19].selectedBy = SelectedBy.PlayerTwo;
+        grid[24].selectedBy = SelectedBy.PlayerTwo;
+        grid[29].selectedBy = SelectedBy.PlayerTwo;
+        grid[34].selectedBy = SelectedBy.PlayerTwo;
+        expect(calculateWinner(grid, isPlayerOne)).toEqual("PLAYERTWO (vertical) Four in a row!");
+    });
+
+    test("PlayerOne descending diagonal win", () => {
+        let grid = createGrid();
+        let isPlayerOne = true;
+        grid[11].selectedBy = SelectedBy.PlayerOne;
+        grid[17].selectedBy = SelectedBy.PlayerOne;
+        grid[23].selectedBy = SelectedBy.PlayerOne;
+        grid[29].selectedBy = SelectedBy.PlayerOne;
+        expect(calculateWinner(grid, isPlayerOne)).toEqual("PLAYERONE (descending diagonal) Four in a row!");
+    });
+    test("PlayerTwo descending diagonal win", () => {
+        let grid = createGrid();
+        let isPlayerOne = false;
+        grid[1].selectedBy = SelectedBy.PlayerTwo;
+        grid[7].selectedBy = SelectedBy.PlayerTwo;
+        grid[13].selectedBy = SelectedBy.PlayerTwo;
+        grid[19].selectedBy = SelectedBy.PlayerTwo;
+        expect(calculateWinner(grid, isPlayerOne)).toEqual("PLAYERTWO (descending diagonal) Four in a row!");
+    });
+
+    test("PlayerOne ascending diagonal win", () => {
+        let grid = createGrid();
+        let isPlayerOne = true;
+        grid[31].selectedBy = SelectedBy.PlayerOne;
+        grid[27].selectedBy = SelectedBy.PlayerOne;
+        grid[23].selectedBy = SelectedBy.PlayerOne;
+        grid[19].selectedBy = SelectedBy.PlayerOne;
+        expect(calculateWinner(grid, isPlayerOne)).toEqual("PLAYERONE (ascending diagonal) Four in a row!");
+    });
+    test("PlayerTwo ascending diagonal win", () => {
+        let grid = createGrid();
+        let isPlayerOne = false;
+        grid[16].selectedBy = SelectedBy.PlayerTwo;
+        grid[12].selectedBy = SelectedBy.PlayerTwo;
+        grid[8].selectedBy = SelectedBy.PlayerTwo;
+        grid[4].selectedBy = SelectedBy.PlayerTwo;
+        expect(calculateWinner(grid, isPlayerOne)).toEqual("PLAYERTWO (ascending diagonal) Four in a row!");
+    });
+
+    test("No win with mixed selected blocks", () => {
+        let grid = createGrid();
+        let isPlayerOne = true;
+        grid[11].selectedBy = SelectedBy.PlayerOne;
+        grid[17].selectedBy = SelectedBy.PlayerTwo;
+        grid[23].selectedBy = SelectedBy.PlayerTwo;
+        grid[29].selectedBy = SelectedBy.Unselected;
+        expect(calculateWinner(grid, isPlayerOne)).toEqual("No winner");
+    });
+    test("Draw when there's no winner & all blocks are selected", () => {
+        let grid = createGrid();
+        let isPlayerOne = true;
+        grid[0].selectedBy = SelectedBy.PlayerOne
```

Comment:
I wonder if you could have made this cleaner using an array of numbers and mapping through the initial grid.

You might need to have an array for each player.

Reply:
Yeah I thought the same thing, but I'll leave this for when I come back to it after the move to Redux